library/src/main/java/stanwood/framework/crm/ApiService.java

```diff
@@ -0,0 +1,91 @@
+package stanwood.framework.crm;
+
+import android.app.Application;
+import android.util.Log;
+
+import java.util.HashMap;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.logging.HttpLoggingInterceptor;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+import stanwood.framework.crm.utils.NetworkService;
+
+
+public class ApiService {
+
+
+    private final Application application;
+    private final CrmApiInterface crmApiInterface;
+    private final Crm crm;
+    private final NetworkService networkService;
+
+    ApiService(Crm crm, Application application) {
+        this.application = application;
+        this.crm = crm;
+
+        OkHttpClient.Builder okHttpBuilder = new OkHttpClient().newBuilder();
+        okHttpBuilder.addInterceptor(chain -> {
+            Request request = chain
+                    .request()
+                    .newBuilder()
+                    .addHeader("Authorization", application.getString(R.string.crm_api_token))
+                    .build();
+
+            return chain.proceed(request);
+        });
+
+        if (BuildConfig.DEBUG) {
+            okHttpBuilder.addInterceptor(
+                    new HttpLoggingInterceptor()
+                            .setLevel(HttpLoggingInterceptor.Level.BODY));
+        }
+
+
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(application.getString(R.string.crm_base_api_url))
+                .client(okHttpBuilder.build())
+                .addConverterFactory(GsonConverterFactory.create())
+                .build();
+
+        crmApiInterface = retrofit.create(CrmApiInterface.class);
+        networkService = new NetworkService(this, application);
+    }
+
+
+    void updateToken(String pushToken, String userId) {
+        HashMap<String, Object> params = new HashMap<>();
```

Comment:
Haha, no offense, but oh those Hashmaps! :D

Seriously, I would really prefer having these parameters defined directly in the CrmApiInterface (usually you would define a class for that with the parameters in the constructor in case of request bodies). Thus there is no way of forgetting parameters and you can clearly see in the interface how the endpoints look like. E.g. check the `LoginBody` class in PYS.

Reply:
You can not forget those parameters as you need to use updateToken method anyway

Reply:
Please let's settle on using objects where ever possible in the future and Hashmaps only for variable parameters. For now it's okay, to get this thing merged as we need it in SNH.